File: electron/groupSenderKeySession.js
/**
 * SSC libsignal group sender keys — main process (Step 2 / Engine 9).
 * File-backed SenderKeyStore using @signalapp/libsignal-client v0.96.4.
 */

const fs = require('fs');
const path = require('path');
const { randomUUID } = require('crypto');
const { FileJsonStore, resolveUnderRoot } = require('./secureFileStore');
const {
  ProtocolAddress,
  SenderKeyRecord,
  SenderKeyDistributionMessage,
  SenderKeyStore,
  processSenderKeyDistributionMessage,
  groupEncrypt,
  groupDecrypt,
} = require('@signalapp/libsignal-client');

function b64encode(buf) {
  return Buffer.from(buf).toString('base64');
}

function b64decode(str) {
  return new Uint8Array(Buffer.from(str, 'base64'));
}

class SscSenderKeyStore extends SenderKeyStore {
  constructor(store) {
    super();
    this.store = store;
  }

  async saveSenderKey(sender, distributionId, record) {
    const key = `${sender.toString()}:${distributionId}`;
    this.store.data[key] = b64encode(record.serialize());
    this.store.save();
  }

  async getSenderKey(sender, distributionId) {
    const key = `${sender.toString()}:${distributionId}`;
    const raw = this.store.data[key];
    if (!raw) return null;
    return SenderKeyRecord.deserialize(b64decode(raw));
  }
}

class GroupSenderKeySession {
  constructor(userDataPath) {
    this.root = resolveUnderRoot(userDataPath, 'ssc-signal');
    this.meta = { localUserId: null, deviceId: '1' };
    this.metaStore = new FileJsonStore(this.root, 'sender_key_meta.json');
    this.senderKeyStore = new SscSenderKeyStore(new FileJsonStore(this.root, 'sender_keys.json'));
  }

  configure({ deviceId, localUserId } = {}) {
    if (deviceId) this.meta.deviceId = String(deviceId);
    if (localUserId) this.meta.localUserId = localUserId;
  }

  localAddress() {
    const userId = this.meta.localUserId;
    if (!userId) {
      throw new Error('GroupSenderKeySession: localUserId not configured');
    }
    return ProtocolAddress.new(userId, Number(this.meta.deviceId) || 1);
  }

  peerAddress(peerId, deviceId = '1') {
    return ProtocolAddress.new(peerId, Number(deviceId) || 1);
  }

  _groupMeta(groupId) {
    if (!this.metaStore.data.groups) this.metaStore.data.groups = {};
    if (!this.metaStore.data.groups[groupId]) {
      this.metaStore.data.groups[groupId] = {
        distributionId: randomUUID(),
        distributed: false,
      };
      this.metaStore.save();
    }
    return this.metaStore.data.groups[groupId];
  }

  getDistributionState(groupId) {
    const group = this._groupMeta(groupId);
    return {
      distributionId: group.distributionId,
      distributed: Boolean(group.distributed),
    };
  }

  markDistributionSent(groupId) {
    const group = this._groupMeta(groupId);
    group.distributed = true;
    this.metaStore.save();
    return { ok: true };
  }

  async createDistributionMessage(groupId) {
    const group = this._groupMeta(groupId);
    const sender = this.localAddress();
    const dist = await SenderKeyDistributionMessage.create(
      sender,
      group.distributionId,
      this.senderKeyStore
    );
    return {
      distributionId: group.distributionId,
      ciphertext: b64encode(dist.serialize()),
    };
  }

  async processDistribution(senderId, deviceId, ciphertextB64) {
    const sender = this.peerAddress(senderId, deviceId);
    const message = SenderKeyDistributionMessage.deserialize(b64decode(ciphertextB64));
    await processSenderKeyDistributionMessage(sender, message, this.senderKeyStore);
    return { ok: true };
  }

  async encryptGroupPlaintext(groupId, plaintext) {
    const group = this._groupMeta(groupId);
    const sender = this.localAddress();
    const bytes = Buffer.from(String(plaintext), 'utf8');
    const ciphertext = await groupEncrypt(
      sender,
      group.distributionId,
      this.senderKeyStore,
      bytes
    );
    return b64encode(ciphertext.serialize());
  }

  async decryptGroupCiphertext(senderId, deviceId, ciphertextB64) {
    const sender = this.peerAddress(senderId, deviceId);
    const bytes = b64decode(ciphertextB64);
    const plain = await groupDecrypt(sender, this.senderKeyStore, bytes);
    return Buffer.from(plain).toString('utf8');
  }
}

let session = null;

function getGroupSenderKeySession(userDataPath) {
  if (!session) {
    session = new GroupSenderKeySession(userDataPath);
  }
  return session;
}

function wipeGroupSenderKeyData(userDataPath) {
  const root = resolveUnderRoot(userDataPath, 'ssc-signal');
  const errors = [];
  for (const name of ['sender_keys.json', 'sender_key_meta.json']) {
    const file = resolveUnderRoot(root, name);
    if (!fs.existsSync(file)) continue;
    try {
      fs.rmSync(file, { force: true });
    } catch (err) {
      errors.push({ file: name, error: err?.message || String(err) });
      console.error('[ssc] group key wipe failed', name, err?.message || err);
    }
  }
  session = null;
  return errors.length ? { ok: false, errors } : { ok: true };
}

module.exports = {
  GroupSenderKeySession,
  getGroupSenderKeySession,
  wipeGroupSenderKeyData,
};

File: android/app/src/main/java/com/supersecurechat/app/GroupSenderKeySession.kt
package com.supersecurechat.app

import org.json.JSONObject
import org.signal.libsignal.protocol.SignalProtocolAddress
import org.signal.libsignal.protocol.groups.GroupCipher
import org.signal.libsignal.protocol.groups.GroupSessionBuilder
import org.signal.libsignal.protocol.message.SenderKeyDistributionMessage
import java.io.File
import java.util.UUID

class GroupSenderKeySession(filesDir: File) {
    private val root = File(filesDir, "ssc-signal")
    val meta = SscSessionMeta()
    private val metaStore = SscFileJsonStore(File(root, "sender_key_meta.json"))
    private val senderKeyStore = SscSenderKeyStore(SscFileJsonStore(File(root, "sender_keys.json")))

    fun configure(opts: JSONObject) {
        opts.optString("deviceId", "").takeIf { it.isNotEmpty() }?.let { meta.deviceId = it }
        opts.optString("localUserId", "").takeIf { it.isNotEmpty() }?.let { meta.localUserId = it }
        if (opts.has("localUserId") && opts.isNull("localUserId")) {
            meta.localUserId = null
        }
    }

    private fun localAddress(): SignalProtocolAddress {
        val userId = meta.localUserId ?: "ssc-local"
        return SignalProtocolAddress(userId, meta.deviceId.toIntOrNull() ?: 1)
    }

    private fun peerAddress(peerId: String, deviceId: String = "1"): SignalProtocolAddress {
        return SignalProtocolAddress(peerId, deviceId.toIntOrNull() ?: 1)
    }

    private fun groupMeta(groupId: String): JSONObject {
        if (!metaStore.data.has("groups")) {
            metaStore.data.put("groups", JSONObject())
        }
        val groups = metaStore.data.getJSONObject("groups")
        if (!groups.has(groupId)) {
            groups.put(
                groupId,
                JSONObject()
                    .put("distributionId", UUID.randomUUID().toString())
                    .put("distributed", false),
            )
            metaStore.save()
        }
        return groups.getJSONObject(groupId)
    }

    fun getDistributionState(groupId: String): JSONObject {
        val group = groupMeta(groupId)
        return JSONObject()
            .put("distributionId", group.getString("distributionId"))
            .put("distributed", group.optBoolean("distributed", false))
    }

    fun markDistributionSent(groupId: String): JSONObject {
        val group = groupMeta(groupId)
        group.put("distributed", true)
        metaStore.save()
        return JSONObject().put("ok", true)
    }

    fun createDistributionMessage(groupId: String): JSONObject {
        val group = groupMeta(groupId)
        val distributionId = UUID.fromString(group.getString("distributionId"))
        val sender = localAddress()
        val dist = GroupSessionBuilder(senderKeyStore).create(sender, distributionId)
        return JSONObject()
            .put("distributionId", group.getString("distributionId"))
            .put("ciphertext", B64.encode(dist.serialize()))
    }

    fun processDistribution(senderId: String, deviceId: String, ciphertextB64: String): JSONObject {
        val sender = peerAddress(senderId, deviceId)
        val message = SenderKeyDistributionMessage(B64.decode(ciphertextB64))
        GroupSessionBuilder(senderKeyStore).process(sender, message)
        return JSONObject().put("ok", true)
    }

    fun encryptGroupPlaintext(groupId: String, plaintext: String): String {
        val group = groupMeta(groupId)
        val distributionId = UUID.fromString(group.getString("distributionId"))
        val sender = localAddress()
        val cipher = GroupCipher(senderKeyStore, sender)
        val message = cipher.encrypt(distributionId, plaintext.toByteArray(Charsets.UTF_8))
        return B64.encode(message.serialize())
    }

    fun decryptGroupCiphertext(senderId: String, deviceId: String, ciphertextB64: String): String {
        val sender = peerAddress(senderId, deviceId)
        val cipher = GroupCipher(senderKeyStore, sender)
        val plain = cipher.decrypt(B64.decode(ciphertextB64))
        return String(plain, Charsets.UTF_8)
    }
}